udemy_enroller/udemy.py

```diff
@@ -25,238 +27,235 @@ class UdemyStatus(Enum):
-            (By.XPATH, enroll_button_xpath)
-        )
-        WebDriverWait(self.driver, 10).until(enroll_button_is_clickable).click()
+        self._checkout(course_id, coupon_code, url)
 
-        # Wait for success page to load
-        success_element_class = "alert-success"
-        WebDriverWait(self.driver, 10).until(
-            EC.presence_of_element_located((By.CLASS_NAME, success_element_class))
-        )
+    def _get_course_id(self, url: str) -> int:
+        response = self.session.get(url)
+        soup = BeautifulSoup(response.content, "html.parser")
 
-        logger.info(f"Successfully enrolled in: {course_name}")
-        return UdemyStatus.ENROLLED.value
+        return int(soup.find("body")["data-clp-course-id"])
 
-    def _check_if_robot(self) -> bool:
+    def _checkout(self, course_id: int, coupon_code: str, url: str) -> str:
+        payload = self._build_checkout_payload(course_id, coupon_code)
+        checkout_result = self.session.post(self.CHECKOUT_URL, json=payload)
+        if not checkout_result.ok:
+            logger.error(
+                f"Checkout failed: Code: {checkout_result.status_code} Text: {checkout_result.text}"
+            )
+            # TODO: Handle non 200 responses
+            #  e.g. Status code: 429 {"detail":"Request was throttled. Expected available in 9 seconds."}
+        else:
+            result = checkout_result.json()
+            if result["status"] == "failed":
+                logger.warning("Checkout failed")
+                # TODO: Shouldn't happen. Need to monitor if it does
+                return UdemyStatus.EXPIRED.value
+            elif result["status"] == "succeeded":
+                logger.info(f"Successfully enrolled: {url}")
+                return UdemyStatus.ENROLLED.value
+
+    @staticmethod
+    def _build_checkout_payload(course_id: int, discount_code: str) -> Dict:
         """
-        Simply checks if the captcha element is present on login if email/password elements are not
 
-        :return: Bool
+        :param course_id:
+        :param discount_code:
+        :return:
         """
-        is_robot = False
-        try:
-            self.driver.find_element_by_id("px-captcha")
-            is_robot = True
-        except NoSuchElementException:
-            pass
-
-        # Check for cloudflare robot check
-        try:
-            self.driver.find_element_by_id("captcha-bypass")
-            is_robot = True
-        except NoSuchElementException:
-            pass
-        return is_robot
+        return {
+            "checkout_event": "Submit",
+            "shopping_cart": {
+                "items": [
+                    {
+                        "discountInfo": {"code": discount_code},
+                        "purchasePrice": {
+                            "amount": 0,
+                            "currency": "EUR",
+                            "price_string": "Free",
+                            "currency_symbol": "€",
```

Comment:
Not sure if this needs to be updated for different regions. Would be good if others could test it out